python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -0,0 +1,238 @@
+import itertools
+from typing import Callable, List, Optional, Tuple
+
+import ray
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.data._internal.split import _split_at_indices
+from ray.data._internal.stats import StatsDict
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+    PhysicalOperator,
+)
+from ray.data.block import (
+    Block,
+    BlockAccessor,
+    BlockExecStats,
+    BlockMetadata,
+    BlockPartition,
+)
+
+
+class ZipOperator(PhysicalOperator):
+    """An operator that zips its inputs together."""
+
+    def __init__(
+        self,
+        input_ops: Tuple[PhysicalOperator, PhysicalOperator],
+    ):
+        """Create a ZipOperator.
+
+        Args:
+            input_ops: The pair of operators to zip together.
+        """
+        self._left_buffer: List[RefBundle] = []
+        self._right_buffer: List[RefBundle] = []
+        self._output_buffer: List[RefBundle] = []
+        self._stats: StatsDict = {}
+        super().__init__("Zip", input_ops)
+
+    def num_outputs_total(self) -> Optional[int]:
+        left_num_outputs = self.input_dependencies[0].num_outputs_total()
+        right_num_outputs = self.input_dependencies[1].num_outputs_total()
+        if left_num_outputs is not None and right_num_outputs is not None:
+            return max(left_num_outputs, right_num_outputs)
+        elif left_num_outputs is not None:
+            return left_num_outputs
+        else:
+            return right_num_outputs
+
+    def add_input(self, refs: RefBundle, input_index: int) -> None:
+        assert not self.completed()
+        assert input_index == 0 or input_index == 1, input_index
+        if input_index == 0:
+            self._left_buffer.append(refs)
+        else:
+            self._right_buffer.append(refs)
+
+    def inputs_done(self) -> None:
+        self._output_buffer, self._stats = self._zip(
+            self._left_buffer, self._right_buffer
+        )
+        self._left_buffer.clear()
+        self._right_buffer.clear()
+        super().inputs_done()
+
+    def has_next(self) -> bool:
+        return len(self._output_buffer) > 0
+
+    def get_next(self) -> RefBundle:
+        return self._output_buffer.pop(0)
+
+    def get_stats(self) -> StatsDict:
+        return self._stats
+
+    def get_transformation_fn(self) -> Callable:
+        return self._zip
+
+    def _zip(
```

Comment:
This is copied from `do_zip_all()` - https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/stage_impl.py#L154 .